File: encode_roman_numerals.rb
# http://rosettacode.org/wiki/Roman_numerals/Encode

def roman_numeral(number)
  roman_numeral_string = String.new
  digits_array = []
  number_string = number.to_s
  iterations = number_string.length-1
  zeros_to_add = iterations
  (0..iterations).each do |number_string_index|
    a = number_string[number_string_index]
    zeros_to_add.times do 
      a << "0"
    end
    digits_array << a
    zeros_to_add -= 1
  end
  digits_array.each do |number|
    case
    when number == "5000"
      roman_numeral_string << "MMMMM"
    when number == "4000"
      roman_numeral_string << "MMMM"
    when number == "3000"
      roman_numeral_string << "MMM"    
    when number == "2000"
      roman_numeral_string << "MM"
    when number == "1000"
      roman_numeral_string << "M"
    when number == "900"
      roman_numeral_string << "CM"
    when number == "800"
      roman_numeral_string << "DCCC"
    when number == "700"
      roman_numeral_string << "DCC"
    when number == "600"
      roman_numeral_string << "DC"
    when number == "500"
      roman_numeral_string << "D"
    when number == "400"
      roman_numeral_string << "CD"
    when number == "300"
      roman_numeral_string << "CCC"
    when number == "200"
      roman_numeral_string << "CC"
    when number == "100"
      roman_numeral_string << "C"
    when number == "90"
      roman_numeral_string << "XC"
    when number == "80"
      roman_numeral_string << "LXXX"
    when number == "70"
      roman_numeral_string << "LXX"
    when number == "60"
      roman_numeral_string << "LX"
    when number == "50"
      roman_numeral_string << "L"
    when number == "40"
      roman_numeral_string << "XL"
    when number == "30"
      roman_numeral_string << "XXX"
    when number == "20"
      roman_numeral_string << "XX"
    when number == "10"
      roman_numeral_string << "X"
    when number == "9" 
      roman_numeral_string << "IX"
    when number == "8"
      roman_numeral_string << "VIII"
    when number == "7"
      roman_numeral_string << "VII"
    when number == "6"
      roman_numeral_string << "VI"
    when number == "5"
      roman_numeral_string << "V"
    when number == "4"
      roman_numeral_string << "IV"
    when number == "3"
      roman_numeral_string << "III"
    when number == "2"
      roman_numeral_string << "II"
    when number == "1"
      roman_numeral_string << "I"
    when number == "0"
      next
    end
  end              
  roman_numeral_string
end  

p "1990 -> #{roman_numeral(1990)}"
p "2008 -> #{roman_numeral(2008)}"
p "1666 -> #{roman_numeral(1666)}"

File: encode_roman_numerals.swift
#!/usr/bin/env swift
// http://rosettacode.org/wiki/Roman_numerals/Encode

import Foundation

func split(n: Int) -> [Int] {
  return reverse(Array(String(n))).map { "\($0)".toInt()! }
}

func romanNumeral(n: Int) -> String {
  var romanNumeralString = ""
  var splitN = split(n).map { Double($0) }
  for i in 0..<splitN.count {
    splitN[i] = splitN[i] * pow(10,Double(i))
  }
  for var i = splitN.count - 1; i >= 0; i-- {
    switch splitN[i] {
    case 5000.0:
      romanNumeralString += "MMMMM"
    case 4000.0:
      romanNumeralString += "MMMM"
    case 3000.0:
      romanNumeralString += "MMM"    
    case 2000.0:
      romanNumeralString += "MM"
    case 1000.0:
      romanNumeralString += "M"
    case 900.0:
      romanNumeralString += "CM"
    case 800.0:
      romanNumeralString += "DCCC"
    case 700.0:
      romanNumeralString += "DCC"
    case 600.0:
      romanNumeralString += "DC"
    case 500.0:
      romanNumeralString += "D"
    case 400.0:
      romanNumeralString += "CD"
    case 300.0:
      romanNumeralString += "CCC"
    case 200.0:
      romanNumeralString += "CC"
    case 100.0:
      romanNumeralString += "C"
    case 90.0:
      romanNumeralString += "XC"
    case 80.0:
      romanNumeralString += "LXXX"
    case 70.0:
      romanNumeralString += "LXX"
    case 60.0:
      romanNumeralString += "LX"
    case 50.0:
      romanNumeralString += "L"
    case 40.0:
      romanNumeralString += "XL"
    case 30.0:
      romanNumeralString += "XXX"
    case 20.0:
      romanNumeralString += "XX"
    case 10.0:
      romanNumeralString += "X"
    case 9.0:
      romanNumeralString += "IX"
    case 8.0:
      romanNumeralString += "VIII"
    case 7.0:
      romanNumeralString += "VII"
    case 6.0:
      romanNumeralString += "VI"
    case 5.0:
      romanNumeralString += "V"
    case 4.0:
      romanNumeralString += "IV"
    case 3.0:
      romanNumeralString += "III"
    case 2.0:
      romanNumeralString += "II"
    case 1.0:
      romanNumeralString += "I"
    default:
      romanNumeralString += ""
    }
  }
  return romanNumeralString
}

for y in [1990,2008,1666] {
  println(romanNumeral(y))
}
